src/components/modals/CompleteQuestModal.tsx

```diff
@@ -138,60 +141,257 @@ const RedeemItem = ({ questName }: Omit<QuestDetailProps, "type">) => {
-        {type === "point" && <RedeemPoint questName={questName} />}
-        {type === "item" && <RedeemItem questName={questName} />}
-      </div>
-      {isQuestCompleted ? (
-        <div className="flex flex-col gap-4">
-          <span className="text-xs text-gray-10 font-light">More quests</span>
-          <div className="flex flex-col gap-2">
-            {quests
-              ?.slice(0, MORE_QUESTS_TO_SHOW)
-              ?.map(
-                (
+  const handleCompleteQuest = async () => {
+    const authToken = getAuthToken();
+    const profile = getProfile();
+    const keys = getKeys();
+
+    if (!authToken || authToken.expiresAt < new Date() || !profile || !keys) {
+      toast.error("You must be logged in to complete a quest");
+      router.push("/login");
+      return;
+    }
+
+    setDisplayState(CompleteQuestDisplayState.PROVING);
+
+    const onUpdateProvingState = (
+      provingStateUpdate: QuestProvingStateUpdate
+    ) => {
+      setProvingState((prevProvingState) => {
+        const newProvingState = { ...prevProvingState };
+        if (provingStateUpdate.numRequirementsUpdate) {
+          newProvingState.numRequirementsTotal =
+            provingStateUpdate.numRequirementsUpdate.numRequirementsTotal;
+          newProvingState.numRequirementsProven =
+            provingStateUpdate.numRequirementsUpdate.numRequirementsProven;
+        }
+        if (provingStateUpdate.currentRequirementUpdate) {
+          newProvingState.currentRequirementNumSigsTotal =
+            provingStateUpdate.currentRequirementUpdate.currentRequirementNumSigsTotal;
+          newProvingState.currentRequirementNumSigsProven =
+            provingStateUpdate.currentRequirementUpdate.currentRequirementNumSigsProven;
+        }
+
+        return newProvingState;
+      });
+    };
+
+    const serializedProof = await generateProofForQuest(
+      quest,
+      onUpdateProvingState
+    );
+
+    const response = await fetch("/api/quest/submit_proof", {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify({
+        questId: quest.id.toString(),
+        authToken: authToken.value,
+        serializedProof,
+      }),
+    });
+
+    if (!response.ok) {
+      toast.error("Failed to submit proof!");
+      setDisplayState(CompleteQuestDisplayState.INITIAL);
+      return;
+    }
+
+    const data = await response.json();
+    if (!data.verified) {
+      toast.error("Proof failed to verify!");
+      setDisplayState(CompleteQuestDisplayState.INITIAL);
+      return;
+    }
+
+    const proofId = data.proofId;
+    if (!proofId) {
+      toast.error("Failed to submit proof!");
+      setDisplayState(CompleteQuestDisplayState.INITIAL);
+      return;
+    }
+
+    const senderPrivateKey = keys.encryptionPrivateKey;
+    const recipientPublicKey = profile.encryptionPublicKey;
+    const encryptedMessage = await encryptQuestCompletedMessage({
```

Comment:
Any reason we don't just combine this with submit proof? I guess we would need to generate a client-side UUID / questID but that seems fine. we could also use hash of the serializedProof. Just trying to minimize number of requests we make and block on, cuz internet is likely to be horrible

Reply:
Also lowers chances of any data loss around completing quests which would be very confusing

Reply:
Added #86. Also created #87, as I was reminded we should not call `loadMessages` separately after sending a jubSignal message.